walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.LinkedList;
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+  private List<TimeRange> busyTimes = new ArrayList<>();
+  private List<TimeRange> freeTimes = new ArrayList<>();
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    busyTimes = getBusyTimes(events, request);
+    if (busyTimes.size() > 0) {
+      return (getFreeTimes(findOverlappingTimes(), request));
+    }
+    return (getFreeTimes(busyTimes, request));
+  }
+
+  // Compare attendees of given events with attendees for requested meeting to determine conflicts
+  private List<TimeRange> getBusyTimes(Collection<Event> events, MeetingRequest request) {
+
+    Collection<String> attendees = request.getAttendees();
+    for (Event event : events) {
+      if(!(Collections.disjoint(event.getAttendees(), attendees))) {
+        busyTimes.add(event.getWhen());
+      }
+    }
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+    return busyTimes;
+  }
+
+  // Returns list of consolidated busy time ranges accounting for overlapping events
+  private List<TimeRange> findOverlappingTimes() {
+    List<TimeRange> overlappingTimes = new ArrayList<>();
+    // Variables used to keep track of start and end times of previous relevant busy time blocks
+    int prevStart = busyTimes.get(0).start();
+    int prevEnd = busyTimes.get(0).end();
+    for (int i = 1; i < busyTimes.size(); i++) {
+      int currentStart = busyTimes.get(i).start();
+      int currentEnd = busyTimes.get(i).end();
+      if (currentStart > prevEnd) {
+        overlappingTimes.add(TimeRange.fromStartEnd(prevStart, prevEnd, false));
+        prevStart = currentStart;
+        prevEnd = currentEnd;
```

Comment:
Using {[4,5], [6,7]} as an example:
prevStart = 4, prevEnd = 5 and currentStart = 6, currentEnd = 7. 
If your currentStart > prevEnd, then the times don't overlap. 

It looks like the logic for this if statement changed since the last commit, so much so that lines 59, and 64 are the same check :). Typo?

Reply:
If we are looking at the same version, line 59 is the false case (from our conversion) where the events don't overlap. Even though this is the false case, I still need to add the events that don't overlap because when working, this function returns a list of all of the consolidated time ranges. Is that what you are addressing?

In terms of line 64, for me, this line of code reads `else if (currentEnd > prevEnd) {` which accounts for the cases where the events overlap but are not nested. In order to get to this else if, the boolean expression on line 59 needs to be false (`if (currentStart > prevEnd) {`), which includes overlapping events. Does this make sense? Do you think that there is a clearer way to express this in the code?